trpcultivate_phenotypes/src/Plugin/Validators/validDataFile.php

```diff
@@ -0,0 +1,197 @@
+   * 
+   * @return array
+   *   An associative array with the following keys.
+   *    - case: a developer focused string describing the case checked.
+   *    - valid: either TRUE or FALSE depending on if the file is valid or not.
+   *    - failedItems: an array of "items" that failed to be used in the message to the user. This is an empty array if the file input was valid.
+   */
+  public function validateFile($filename, $fid = NULL) {
+    // Validator response values for a valid file.
+    $case = 'Data file is valid';
+    $valid = TRUE;
+    $failed_items = [];
+
+    // Parameter check, verify that the filename/file path is valid.
+    if (empty($filename) && is_null($fid)) {
+      return [
+        'case' => 'Filename is empty',
+        'valid' => FALSE,
+        'failedItems' => ['filename' => $filename]
+      ];
+    }
+    
+    // Parameter check, verify the file id number is not 0 or negative values.
+    if (!is_null($fid) && $fid <= 0) {
+      return [
+        'case' => 'Invalid file id number',
+        'valid' => FALSE,
+        'failedItems' => ['file_id' => $fid]
+      ];
+    } 
+
+    // File input.
+    $file_input = (is_null($fid)) ? $filename : $fid;
+    
+    // Load file object.
+    if (is_numeric($file_input)) {
+      // The file input is integer value, the file id number.
+      // Load the file object by fid number.
+      $file_id = $fid;
+      $file_object = File::load($file_id);
+
+      $input_key = 'file_id';
+    }
+    else {
+      // The file input is a string value, a path to the file.
+      // Locate the file entity by uri and load the file object using
+      // the returned file id number that matched.
+      $file_entities = $this->service_EntityTypeManager
+        ->getStorage('file')
+        ->loadByProperties(['uri' => $filename]);
+      
+      $file_entity = reset($file_entities);
+      $file_object = 0;
+      
+      if ($file_entity) {
+        $file_id = $file_entity->get('fid')->value;
+        $file_object = File::load($file_id);
+      }
+
+      $input_key = 'filename';
+    }
+    
+
+    // Validate data file:
+
+    if ($file_object) {
+      // File object has loaded successfully.
+      
+      // Check that the file is not blank by inspecting the file size
+      // to see if it is greater than 0.
+      $file_size = $file_object->getSize();
+      if ($file_size > 0) {
+      
+        // Check that both the file extension and file MIME type
+        // are supported by the importer.
+        $file_mime_type = $file_object->getMimeType();
+        $file_filename  = $file_object->getFileName();
+        $file_extension = pathinfo($file_filename, PATHINFO_EXTENSION);
+
+        $supported_file_extensions = $this->getSupportedFileExtensions();
+        $supported_mime_types = $this->getSupportedMimeTypes();
+        if (in_array($file_mime_type, $supported_mime_types) && in_array($file_extension, $supported_file_extensions)) {
```

Comment:
I think these should be separate checks to do more helpful error reporting. However, I wonder if file extension should be checked for at all here. If anything, it can be used in the case string. For example:
`$case = 'The file type is not supported by this importer. Make sure your file is one of the following: $supported_file_extensions'`

Reply:
I agree with separating these. Also remember these are developer messages.

Check 1: is the mime type in the supported mime types?
  if not then...
    check file extension is one of supported?
      if not then failed case 'unsupported file mime/extension', failedItems includes both the mime-type and file extension of file entity.
      if yes then failed case 'unsupported file mime and mismatched extension', failedItems includes both the mime-type and file extension of file entity.
  if yes... all is well, we don't care about file extension.

Reply:
My concerns here are resolved ✅